src/Microsoft.OData.Core/ObjectPool.cs

```diff
@@ -0,0 +1,100 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ObjectPool.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// Manages a pool of reusable objects. An object retrieved from the pool should be returned when it's no longer being used so that it can be later recycled.
+    /// </summary>
+    /// <typeparam name="T">The type of objects to pool.</typeparam>
+    /// <remarks>This object pool is not thread-safe and not intended to be shared across concurrent requests.
+    /// Consider using the DefaultObjectPool in 8.x once we drop support for netstandard 1.1
+    /// </remarks>
+    internal class ObjectPool<T> where T : class
+    {
+        private readonly Func<T> objectGenerator;
+        private ObjectWrapper[] items;
+        private T firstItem;

```

Comment:
I'm not sure what benefit we get out of keeping track of the first item instead of keeping track of the index of the first item (maybe we are trying to remove one deference as an optimization?). Doing this without some other field means that we *must* use `null` to keep track of what slots in `items` are available, which means that `T` *must* be constrained to reference types. We also can't tell the difference between a pool that has loaned all of its objects, or a pool that hasn't generated any objects. This means that we won't notice when we have introduced a 9th level of nesting in a response, and we will start hitting perf issues at that point because the `Get` method will see `items` as only containing `null` values, so it will call `objectGenerator`, and the `Return` method will see `items` as being completely full and drop the returned object on the floor.

Reply:
I think avoiding array allocations on responses with 1 level of nesting could be beneficial.  I don't think it's a problem to constrain `T` to reference types since we are trying to reduce heap allocations. If we were storing value types in the pool, new copies would be created when we call `Get()` anyway.

I think you make a good point about the hitting the 9th level of nesting. Though I don't understand what you meant by "`Return` method will see `items` as being completely full and drop the returned object on the floor". If `Return` sees `items` as full, it will replace one of the `null` slots with the items being returned.

What would you propose `Get()` do when hitting the max pool size? I feel like throwing an exception is too extreme.

Reply:
I have modified the code:
- When an object cannot be returned to the array when the array is full, we dispose it
- When we are done with the object pool, it's disposed

Reply:
I'm not sure why we ended up implementing `IDisposable` here, but I think this will ultimately be resolved by understanding what we really want the implementation to do (see the other open comment).